core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1080,6 +1080,21 @@ abstract class RDD[T: ClassTag](
       var partiallyAggregated = mapPartitions(it => Iterator(aggregatePartition(it)))
       var numPartitions = partiallyAggregated.partitions.length
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
+
+      // Do one level of aggregation based on executorId before starting the tree
+      // NOTE: exclude the driver from list of executors
+      val numExecutors = math.max(context.getExecutorStorageStatus.length - 1, 1)
+      partiallyAggregated = partiallyAggregated.mapPartitionsWithIndex { case (idx, iter) =>
+        def isAllDigits(x: String) = x forall Character.isDigit
+        val execId = SparkEnv.get.executorId
+        if (isAllDigits(execId)) {
+          iter.map((execId.toInt, _))
+        } else {
+          iter.map((execId.hashCode, _))
```

Comment:
Ah I can actually reproduce it -- so the problem is that the keyspace get un-balanced when we use hashCode on Strings. For example if you start with 100 exec ids and want to get 50 partitions out, some partitions will get 3 ids while some will get 1
```
(0 until 100).map(_.toString).map(_.hashCode % 50).groupBy(identity).map(_._2.size)
```
while with integers you get a uniform distribution
```
(0 until 100).map(_.hashCode % 50).groupBy(identity).map(_._2.size)
```

So you could end up with unbalanced trees which is not desirable. 

@andrewor14 Can we guarantee that execId will be an integer ? Is there any other integer we can get a handle on otherwise ?